composer/datasets/cifar.py

```diff
@@ -49,10 +68,84 @@ def initialize_object(self, batch_size: int, dataloader_hparams: DataLoaderHpara
+                from ffcv.pipeline.operation import Operation  # type: ignore
+            except ImportError:
+                raise ImportError(
+                    textwrap.dedent("""\
+                    Composer was installed without ffcv support.
+                    To use ffcv with Composer, run `pip install mosaicml[ffcv]`"""))
+
+            dataset_file = self.ffcv_dest_train if self.is_train else self.ffcv_dest_val
+            dataset_filepath = os.path.join(self.ffcv_dir, dataset_file)
+            # always create if ffcv_write_dataset is true
+            if self.ffcv_write_dataset and not os.path.exists(dataset_filepath):
+                log.info(f"Writing dataset in FFCV <file>.beton format to {dataset_filepath}.")
+                if self.datadir is None:
+                    raise ValueError("datadir is required if use_synthetic is False and ffcv_write_dataset is True.")
+                ds = CIFAR10(
+                    self.datadir,
+                    train=self.is_train,
+                    download=self.download,
+                )
+
+                writer = ffcv.writer.DatasetWriter(dataset_filepath, {
+                    'image': ffcv.fields.RGBImageField(),
+                    'label': ffcv.fields.IntField()
+                })
+                writer.from_indexed_dataset(ds)
+
+            if not os.path.exists(dataset_filepath):
+                raise ValueError(
+                    f"Dataset file containing samples not found at {dataset_filepath}. Use ffcv_dir flag to point to a dir containing {dataset_filepath}."
+                )
+
+            # Please note that this mean/std is different from the mean/std used for regular PyTorch dataloader as
+            # ToTensor does the normalization for PyTorch dataloaders.
+            cifar10_mean_ffcv = [125.307, 122.961, 113.8575]
+            cifar10_std_ffcv = [51.5865, 50.847, 51.255]
+            label_pipeline: List[Operation] = [IntDecoder(), ffcv.transforms.ToTensor(), ffcv.transforms.Squeeze()]
+            image_pipeline: List[Operation] = [SimpleRGBImageDecoder()]
+
+            if self.is_train:
+                image_pipeline.extend([
+                    ffcv.transforms.RandomHorizontalFlip(),
+                    ffcv.transforms.RandomTranslate(padding=2, fill=tuple(map(int, cifar10_mean_ffcv))),
+                    ffcv.transforms.Cutout(4, tuple(map(int, cifar10_mean_ffcv))),
+                ])
+            # Common transforms for train and test
+            image_pipeline.extend([
+                ffcv.transforms.ToTensor(),
+                ffcv.transforms.ToTorchImage(channels_last=False, convert_back_int16=False),
+                ffcv.transforms.Convert(torch.float32),
+                transforms.Normalize(cifar10_mean_ffcv, cifar10_std_ffcv),
+            ])
+
+            ordering = ffcv.loader.OrderOption.RANDOM if self.is_train else ffcv.loader.OrderOption.SEQUENTIAL
+
+            return ffcv.Loader(
```

Comment:
surprised this doesn't raise a pyright error....

Reply:
:-)